### DISCUSSION TITLE
How to use Desktop object to connect to app with changing window titles

### DISCUSSION THREAD

**Discussion by ctechn** [pywinauto/pywinauto#1259] [Q&A]
@vasily-v-ryabov hoping you can help here. I originally posted to Stackoverflow but no one was able to assist successfully.

 I am trying to automate the Spotify app on Windows (i.e. the Microsoft Store option). I am using UIA backend. The only problem I am having is that because pywinauto's Desktop object is looking for a specific window title, it doesn't allow me to automate as Spotify's window title can change depending on what song is being played at any given moment (and there's no word "Spotify" in the window title).

I experimented with trying to obtain the process ID, but it doesn't seem to find the exe name (Spotify.exe) or doesn't work with the Desktop object, which I need to be able to access child_window, etc. My code is below.

Can someone assist with advising how to go about making the Desktop object work when the app's window title is not constant and can't be used to identify the app, and since it's a Microsoft Store app, it spawns multiple processes so using just the Application object is not an option either.

This is my code:

```
import pywinauto
import subprocess
from pywinauto import Desktop, application
from pywinauto.application import Application

sp = subprocess.call("start shell:AppsFolder\SpotifyAB.SpotifyMusic_zpdnekdrzrea0!Spotify",shell=True)
app = pywinauto.Application(backend='uia').connect(path="Spotify.exe")

#this works only if there is NO song playing on Spotify; if any song is playing, 
#it fails to work since it can't find any app window with the title "Spotify"
#and the name of the window is instead the current song playing
#which I don't know what it is going to be at any given point
dlg = Desktop(backend="uia").Spotify
#does not work
dlg = Desktop(backend="uia").window(process=app.process)

dlg.child_window(title="Your Library", control_type="Hyperlink",found_index=0).wait("exists",30).invoke()

dlg.child_window(title="Orchestra", control_type="Hyperlink",found_index=0).wait("exists",30).invoke()
try:
    dlg.child_window(title="Play Orchestra", control_type="Button").wait("exists",5).invoke()
except:
    pass

    # this also doesn't work most of the time
    dlg.wait("",10).close()
```

I tried using what someone suggested (`dlg = app.top_window()`) along with `pid = application.process_from_module(module = 'Spotify.exe')  app = Application(backend='uia').connect(process = pid)` but those remedies didn't work. (Error is still `RuntimeError("No windows for that process could be found")`. Spotify's store app spawns multiple processes which probably explain the issue and with changing window titles, it seems impossible to be able to reference Spotify altogether.

**Comment by ctechn** [comment]
Can anyone please help with this? Still stuck on getting this to work!

**Comment by ctechn** [comment]
@vasily-v-ryabov could you please assist in any way? Thank you

**Comment by vasily-v-ryabov** [comment]
There is possible keyword `title_re` for both `connect`, `window` and `child_window` methods. Also method `connect` has known issue: you have to use `timeout=5` explicitly while connecting by title or title_re. Regular expression should work the same way as from `re` module.

**Reply by vasily-v-ryabov** [reply]
Example: `title_re=".* - Spotify$"` where `.*` means any symbols, `$` means end of string (no more symbols after word "Spotify").

**Reply by ctechn** [reply]
Thanks for the help! Unfortunately, the word "Spotify" is nowhere to be found in the window title for the app, as it only displays the artist and title of the currently playing song (unless paused, which we can't control and want to be able to access the window regardless). So there is another objective way to reference the app, maybe by icon (as that is constant) or the process(es) it spawns? `dlg = Desktop(backend="uia").window(title_re=".*Spotify.*")` or `app = pywinauto.Application(backend='uia').connect(title_re=".*Spotify.*",timeout=5)` does not work, so not sure if there is another way?

**Reply by vasily-v-ryabov** [reply]
Oh, that's a good question. Unfortunately I don't have Spotify on my desktop, and sanctions will not allow it. Screenshot with `Inspect.exe` properties of main window would be potentially helpful.

**Reply by ctechn** [reply]
Unfortunately, no useful data it seems, except for AutomationID? Name is set to a song I was playing.

![Untitled](https://user-images.githubusercontent.com/94086345/216846829-fd8e29ce-181f-4257-925a-6073e809779e.jpg)

**Comment by vasily-v-ryabov** [comment]
Oh, great! But AFAIK `auto_id_re` keyword is supported in `atspi` branch only, not in current release. So at least you can try:
```
pip uninstall -y pywinauto
pip install https://github.com/pywinauto/pywinauto/archive/refs/heads/atspi.zip
```

**Reply by vasily-v-ryabov** [reply]
Also please switch `Inspect.exe` from MSAA to "UI Automation" mode in the future.

**Reply by vasily-v-ryabov** [reply]
Also maybe `title_re=".*running window"` would help with current release.
